### DISCUSSION TITLE
How to use SQLAlchemy under FastAPI?

### DISCUSSION THREAD

**Discussion by Chise1** [sqlalchemy/sqlalchemy#5837] [Usage Questions] [answered]
How to open connect befare server start and close befare server stop?
Can I use it like this?

`

app = FastAPI()
print(settings.DATABASE_URL)
engine = create_async_engine(
    settings.DATABASE_URL, echo=True, future=True
)
session = AsyncSession(engine, future=True)

@app.on_event("startup")
async def startup():
    await session.begin()

@app.on_event("shutdown")
async def shutdown():
    await session.close_all()

`

**Comment by PhillCli** [comment, marked as answer]
@Chise1
you can define dependency that provides the session and use it within the handlers.

```Python
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine  # noqa

engine = create_async_engine("")

SessionLocal = sessionmaker(
    class_=AsyncSession,
    autocommit=False,
    autoflush=False,
    bind=engine
)

async def get_db_session() -> AsyncSession:
    async with SessionLocal() as session:
        yield session

@app.route("/")
async def sample_handler(
    db_session: AsyncSession = Depends(get_db_session),
):
    pass
````

EDIT: added imports, but dude you can create it any way you want, just pointed that `Dependency()` is the way to go.

**Reply by iasukas** [reply]
Thank you for your answer. I have a question here. How to use DB_SESSION to complete a SELECT in sample_handler function?

The query for 1.3 looks like this
`session.query(Student).filter(Student.name == 'Frank').first()`

The 1.4 AsyncSession version does not have a query method, so I use the following method to query
`db.execute(select(Student).where(Student.name== 'Frank')).first()`

But there is no first method for reporting an error
So how do I complete such a query? 
Thank you very much

**Reply by PhillCli** [reply]
refer to [docs](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#synopsis-orm)

I belive you're missing `await`
```python
(await db.execute(select(Student).where(Student.name== 'Frank'))).first()
```
EDIT:
 see [AsyncResult API](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncResult.first) (that's what you get after awaiting session.execute)

**Comment by Chise1** [comment]
Thank you!

**Comment by Chise1** [comment]
Where is SessionLocal ? 
I can't found it in document or package.

**Comment by Chise1** [comment]
Please give me a complete example.

**Comment by jvanasco** [comment]
@Chise1 This is really a question and request for the FastAPI project, not SQLAlchemy. I recommend you ask the question in one of their groups, and update this Issue with the solution for the benefit of others.
